packages/flutter/lib/src/cupertino/sliding_segmented_control.dart

```diff
@@ -0,0 +1,1063 @@
+    }
+    if (newValue != _highlighted && newValue != null) {
+      _pressControllers[newValue].animateTo(1, duration: _kOpacityAnimationDuration, curve: Curves.ease);
+    }
+    _pressed = newValue;
+  }
+
+  void didChangeSelectedViaGesture() {
+    controller.value = _highlighted;
+  }
+
+  T indexToKey(int index) => index == null ? null : keys[index];
+
+  @override
+  Widget build(BuildContext context) {
+    switch (_textDirection) {
+      case TextDirection.ltr:
+        keys = widget.children.keys.toList(growable: false);
+        break;
+      case TextDirection.rtl:
+        keys = widget.children.keys.toList().reversed.toList(growable: false);
+        break;
+    }
+
+    final List<Widget> children = <Widget>[];
+    for (T currentKey in keys) {
+      final TextStyle textStyle = DefaultTextStyle.of(context).style.copyWith(
+        fontWeight: _highlightTween.evaluate(_highlightControllers[currentKey]),
+      );
+
+      final Widget child = DefaultTextStyle(
+        style: textStyle,
+        child: Semantics(
+          button: true,
+          onTap: () { controller.value = currentKey; },
+          inMutuallyExclusiveGroup: true,
+          selected: controller.value == currentKey,
+          child: Opacity(
+            opacity: _pressTween.evaluate(_pressControllers[currentKey]),
+            // Expand the hitTest area to be as large as the Opacity widget.
+            child: MetaData(
+              behavior: HitTestBehavior.opaque,
+              child: Center(child: widget.children[currentKey]),
+            ),
+          ),
+        ),
+      );
+
+      children.add(child);
+    }
+
+    final int selectedIndex = controller.value == null ? null : keys.indexOf(controller.value);
+
+    final Widget box = _SegmentedControlRenderWidget<T>(
+      children: children,
+      selectedIndex: selectedIndex,
+      thumbColor: widget.thumbColor,
+      state: this,
+    );
+
+    return UnconstrainedBox(
+      constrainedAxis: Axis.horizontal,
+      child: Container(
+        padding: widget.padding.resolve(_textDirection),
+        decoration: BoxDecoration(
+          borderRadius: const BorderRadius.all(Radius.circular(_kCornerRadius)),
+          color: CupertinoDynamicColor.resolve(widget.backgroundColor, context),
+        ),
+        child: UnconstrainedBox(constrainedAxis: Axis.horizontal, child: box),
+      ),
+    );
+  }
+}
+
+class _SegmentedControlRenderWidget<T> extends MultiChildRenderObjectWidget {
+  _SegmentedControlRenderWidget({
+    Key key,
+    List<Widget> children = const <Widget>[],
+    @required this.selectedIndex,
+    @required this.thumbColor,
+    @required this.state,
+  }) : super(key: key, children: children);
+
+  final int selectedIndex;
+  final Color thumbColor;
+  final _SegmentedControlState<T> state;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderSegmentedControl<T>(
+      selectedIndex: selectedIndex,
```

Comment:
On the face of it, it seems odd to pass along selectedIndex and the widget state (which could provide a selectedIndex getter) but not update selectedIndex in updateRenderObject.

Reply:
`selectedIndex` is conditionally updated in `updateRenderObject`. If the user is currently dragging the thumb around attempting to update `selectedIndex` programmatically will have no effect.
```
      ..guardedSetHighlightedIndex(selectedIndex);
```